### ISSUE TITLE
optional protocols being limited to classes that inherit from Object-C classes: compiler check?

### ISSUE THREAD

**Issue by ByunghoonKim** [apple/swift#61902]
**Describe the bug**

Hi, I am currently learning Swift programming language.
[Optional Protocol Requirements](https://docs.swift.org/swift-book/LanguageGuide/Protocols.html#ID284l) section in the official Swift programming language document(Swift 5.7) states

> Note that @objc protocols can be adopted only by classes that inherit from Objective-C classes or other @objc classes. They can’t be adopted by structures or enumerations.

and provides an example code.
```
@objc protocol CounterDataSource {
    @objc optional func increment(forCount count: Int) -> Int
    @objc optional var fixedIncrement: Int { get }
}

// ...

class ThreeSource: NSObject, CounterDataSource {
    let fixedIncrement = 3
}
```

Xcode, however, readily accepts the following code snippet which does *not* inherit NSObject, and compiles without any problem.
```
class ThreeSource: /*NSObject,*/ CounterDataSource {
    let fixedIncrement = 3
}
```

Then, am I to understand that the document's description on the limitation is about why optional protocol is introduced to the language in the first place, and not about what is actually enforced by the compiler?

**Screenshots**
<img width="1142" alt="Screenshot 2022-11-03 at 17 50 45" src="https://user-images.githubusercontent.com/44054076/199681061-752da893-9a72-428e-bcbd-9c2c7b641b1f.png">

**Environment (please fill out the following information)**
 - OS: [macOS 13.0]
 - Xcode Version/Tag/Branch: Xcode Version 14.1 (14B47b)

**Comment by AnthonyLatsis** [comment]
@slavapestov Is this a bug? I thought `class ThreeSource` would be implicitly marked as `@objc`, but it isn’t.

**Comment by slavapestov** [comment]
Non-@objc classes can conform to @objc protocols. The documentation is incorrect here.
